shell/platform/tizen/tizen_vsync_waiter.cc

```diff
@@ -4,37 +4,102 @@
 
 #include "tizen_vsync_waiter.h"
 
+#include <eina_thread_queue.h>
+
 #include "flutter/shell/platform/tizen/tizen_embedder_engine.h"
 #include "flutter/shell/platform/tizen/tizen_log.h"
 
-void TizenVsyncWaiter::RequestVblank(void* data, Ecore_Thread* thread) {
-  TizenVsyncWaiter* tizen_vsync_waiter =
-      reinterpret_cast<TizenVsyncWaiter*>(data);
-  if (!ecore_thread_check(thread)) {
-    tdm_error error = tdm_client_vblank_wait(tizen_vsync_waiter->vblank_, 1,
-                                             TdmClientVblankCallback, data);
-    if (error != TDM_ERROR_NONE) {
-      FT_LOGE("tdm_client_vblank_wait error  %d", error);
-      return;
-    }
-    tdm_client_handle_events(tizen_vsync_waiter->client_);
-  }
-}
+static const int QUEUE_QUIT = -1;
+static const int QUEUE_REQUEST_VBLANK = 0;
+
+typedef struct {
+  Eina_Thread_Queue_Msg head;
+  int value;
+} Msg;
+
+static Eina_Thread_Queue* vblank_thread_queue{nullptr};
 
 TizenVsyncWaiter::TizenVsyncWaiter(TizenEmbedderEngine* engine)
     : engine_(engine) {
   if (!CreateTDMVblank()) {
     FT_LOGE("Failed to create TDM vblank");
     DestoryTDMVblank();
+  } else {
+    vblank_thread_queue = eina_thread_queue_new();
+    vblank_thread_ =
+        ecore_thread_feedback_run(RequestVblankLoop, NULL, VblankLoopFinish,
+                                  VblankLoopFinish, this, EINA_TRUE);
   }
 }
 
-TizenVsyncWaiter::~TizenVsyncWaiter() { DestoryTDMVblank(); }
+TizenVsyncWaiter::~TizenVsyncWaiter() {
+  SendMessage(QUEUE_QUIT);
+  if (vblank_thread_) {
+    ecore_thread_cancel(vblank_thread_);
+    vblank_thread_ = nullptr;
+  }
+  DestoryTDMVblank();
+}
 
 void TizenVsyncWaiter::AsyncWaitForVsync(intptr_t baton) {
   baton_ = baton;
   if (TDMValid()) {
-    ecore_thread_run(RequestVblank, NULL, NULL, this);
+    SendMessage(QUEUE_REQUEST_VBLANK);
+  }
+}
+
+void TizenVsyncWaiter::SendMessage(int val) {
+  if (!vblank_thread_queue || !vblank_thread_) {
+    FT_LOGE("vblank thread or vblank thread queue not valid");
+    return;
+  }
+  Msg* msg;
+  void* ref;
+  msg = (Msg*)eina_thread_queue_send(vblank_thread_queue, sizeof(Msg), &ref);
+  msg->value = val;
+  eina_thread_queue_send_done(vblank_thread_queue, ref);
+}
+
+void TizenVsyncWaiter::RequestVblankLoop(void* data, Ecore_Thread* thread) {
+  TizenVsyncWaiter* tizen_vsync_waiter =
+      reinterpret_cast<TizenVsyncWaiter*>(data);
+  void* ref;
+  Msg* msg;
+  while (!ecore_thread_check(thread)) {
+    if (!vblank_thread_queue) {
+      FT_LOGE("Vblank thread queue is not valid");
+      return;
+    }
+    msg = (Msg*)eina_thread_queue_wait(vblank_thread_queue, &ref);
+    if (msg) {
+      eina_thread_queue_wait_done(vblank_thread_queue, ref);
+    } else {
+      FT_LOGE("Message is null");
+      continue;
+    }
+    if (msg->value == QUEUE_QUIT) {
+      FT_LOGD("Message queue quit");
```

Comment:
Is there anything else we need to do other than logging when received a QUEUE_QUIT

Reply:
we don't need to do anything when  send QUEUE_QUIT.  send QUEUE_QUIT  just for exiting the loop, make thread not blocking.